src/GraphQL/Instrumentation/FieldMiddlewareBuilder.cs

```diff
@@ -12,77 +9,36 @@ namespace GraphQL.Instrumentation
     /// </summary>
     public class FieldMiddlewareBuilder : IFieldMiddlewareBuilder
     {
-        private Func<ISchema, FieldMiddlewareDelegate, FieldMiddlewareDelegate> _singleMiddleware;
-        private IList<Func<ISchema, FieldMiddlewareDelegate, FieldMiddlewareDelegate>> _middlewares;
+        private Func<FieldMiddlewareDelegate, FieldMiddlewareDelegate> _middleware;
 
         /// <inheritdoc/>
-        public IFieldMiddlewareBuilder Use(Func<ISchema, FieldMiddlewareDelegate, FieldMiddlewareDelegate> middleware)
+        public IFieldMiddlewareBuilder Use(Func<FieldMiddlewareDelegate, FieldMiddlewareDelegate> middleware)
         {
-            middleware = middleware ?? throw new ArgumentNullException(nameof(middleware));
+            if (middleware == null)
+                throw new ArgumentNullException(nameof(middleware));
 
-            if (_singleMiddleware == null)
+            if (_middleware == null)
             {
-                // allocation free optimization for single middleware (InstrumentFieldsMiddleware)
-                _singleMiddleware = middleware;
-            }
-            else if (_middlewares == null)
-            {
-                _middlewares = new List<Func<ISchema, FieldMiddlewareDelegate, FieldMiddlewareDelegate>>
-                {
-                    _singleMiddleware,
-                    middleware
-                };
+                _middleware = middleware;
             }
             else
             {
-                _middlewares.Add(middleware);
+                var firstMiddleware = _middleware;
+                _middleware = (next) => firstMiddleware(middleware(next));
```

Comment:
```suggestion
                _middleware = next => firstMiddleware(middleware(next));
```

Thanks for rewriting this builder. For some reason, it did not occur to me to combine middleware immediately after adding them. So I had to iterate over them in reverse order.